onnx/shape_inference/implementation.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include "onnx/proto_utils.h"
+#include "onnx/defs/schema.h"
+
+namespace ONNX_NAMESPACE { namespace shape_inference {
+
+struct InferenceContextImpl : public InferenceContext {
+  const AttributeProto* getAttribute(const std::string& name) const override {
+    auto iter = attributesByName_.find(name);
+    if (iter == attributesByName_.end()) {
+      return nullptr;
+    } else {
+      return iter->second;
+    }
+  }
+  size_t getNumInputTypes() const override {
+    return allInputTypes_.size();
+  }
+  const TypeProto_Tensor* getInputType(size_t index) const override {
+    return allInputTypes_[index];
+  }
+  size_t getNumOutputTypes() const override {
+    return allOutputTypes_.size();
+  }
+  TypeProto_Tensor* getOutputType(size_t index) override {
+    return &allOutputTypes_[index];
+  }
+  std::unordered_map<std::string, const AttributeProto *> attributesByName_;
+  std::vector<const TypeProto_Tensor*> allInputTypes_;
+  std::vector<TypeProto_Tensor> allOutputTypes_;
+};
+
+void InferShapes(ModelProto& m) {
+  auto* g = m.mutable_graph();
+
+  std::unordered_map<std::string, TypeProto_Tensor *> valueTypesByName;
+  for (auto& vi : *g->mutable_value_info()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+  for (auto& vi : *g->mutable_input()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+  for (auto& vi : *g->mutable_output()) {
+    valueTypesByName[vi.name()] = vi.mutable_type()->mutable_tensor_type();
+  }
+
+  for (const auto& n : g->node()) {
+    auto schema = OpSchemaRegistry::Schema(n.op_type());
+    if (schema) {
+      InferenceContextImpl ctx;
+
+      {
+        std::unordered_map<std::string, const AttributeProto *> attributesByName;
+        for (const auto& attr : n.attribute()) {
+          attributesByName[attr.name()] = &attr;
+        }
+        ctx.attributesByName_ = std::move(attributesByName);
+      }
+
+      {
+        std::vector<const TypeProto_Tensor *> allInputTypes;
+
+        for (const auto& input : n.input()) {
+          auto iter = valueTypesByName.find(input);
+          if (iter != valueTypesByName.end()) {
+            allInputTypes.push_back(iter->second);
+          } else {
+            allInputTypes.push_back(nullptr);
+          }
+        }
+
+        ctx.allInputTypes_ = std::move(allInputTypes);
+      }
+
+      ctx.allOutputTypes_.resize(n.output_size());
+
+      schema->GetShapeInferenceFunction()(ctx);
+
+      for (int i = 0; i < n.output_size(); i++) {
+        const auto& output = n.output(i);
+        const auto& inferredType = *ctx.getOutputType(i);
+
+        // In this case, we have no new information, so don't bother
+        // to add a contentless ValueInfo.
+        if (inferredType.elem_type() == TensorProto::UNDEFINED &&
+            !inferredType.has_shape()) {
+          continue;
+        }
+
+        // If there is already a ValueInfo associated with this
```

Comment:
How about let's change the code starts from here to the end into:
```
auto iter = valueTypesByName.find(output);
if (iter != valueTypesByName.end()) {
   assert_equal_value_info(iter->second, inferredType);
} else {
   valueTypesByName[output] = inferredType;
}
```
(of course will need to add a separate function `assert_equal_value_info` dedicated for doing the equality check).

Reply:
it's not an equality check though. There can be an existing value info with partial information, which needs to be merged in (e.g. one of the dims is "x3", and we infer it to be 5

Reply:
oh right, then how about
```
auto iter = valueTypesByName.find(output);
if (iter != valueTypesByName.end()) {
   assert_loose_equal_value_info(iter->second, inferredType);
}
iter->second = inferredType;
```

Reply:
well that also doesn't quite work - the inferredType may also lack information which the existing ValueInfo already contains.

I'll pull the logic out in some fashion though